### PR TITLE
Added currencies exchange rates service.

### PR DESCRIPTION
**Pull request by deadsa5**
Added currencies exchange rates update CronJob with trigger at 15:00 every day.

### CODE REVIEW

File: bin/custom/tyreshopcore/src/com/reply/tyreshop/core/job/TyreshopUpdateCurrencyRatesJob.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    @Resource
+    ConfigurationService configurationService;
+
+    private CurrencyDao currencyDao;
+    private TyreshopBYNRatesService currenciesRatesService;
+
+    @Override
+    public PerformResult perform(CronJobModel cronJobModel) {
+        boolean success = true;
+        List<CurrencyModel> currencies = currencyDao.findCurrencies();
```
Comment (uhpavel) [inline bin/custom/tyreshopcore/src/com/reply/tyreshop/core/job/TyreshopUpdateCurrencyRatesJob.java]: Please refactor:
- use one call http://www.nbrb.by/API/ExRates/Rates?Periodicity=0
- retrieve all currencies rates
- update currencies in hybris

File: bin/custom/tyreshopcore/resources/tyreshopcore-spring.xml [hunk 2]
```diff
@@ -562,4 +562,16 @@
 	<bean id="tyreshopCategoryNameValueProvider" parent="abstractCategoryNameValueProvider">
 		<property name="categorySource" ref="tyreshopCategorySource" />
 	</bean>
+
+	<alias name="defaultTyreshopBYNRatesService" alias="tyreshopBYNRatesService" />
+	<bean id="defaultTyreshopBYNRatesService"
```
Comment (uhpavel) [inline bin/custom/tyreshopcore/resources/tyreshopcore-spring.xml]: remove BYN in class name. ExchangeRateService

File: bin/custom/tyreshopcore/resources/tyreshopcore/import/common/essential-data.impex [hunk 3]
```diff
@@ -108,4 +108,10 @@ INSERT_UPDATE ServicelayerJob;code[unique=true];springId[unique=true]
...
 UPDATE SearchRestriction;code[unique=true];active[default=false]
-;Frontend_RestrictedCategory
\ No newline at end of file
+;Frontend_RestrictedCategory
+
+INSERT_UPDATE CronJob;code[unique=true];job(code);singleExecutable;sessionLanguage(isocode)
+                     ;tyreshopUpdateCurrencyRatesCronJob;tyreshopUpdateCurrencyRatesJob;false;en
+
+INSERT_UPDATE Trigger;cronJob(code)[unique=true];cronExpression
+                     ;tyreshopUpdateCurrencyRatesCronJob;0 0 15 * * ?
```
Comment (uhpavel) [inline bin/custom/tyreshopcore/resources/tyreshopcore/import/common/essential-data.impex]: trigger at the start of a day to retrieve currencies for today

File: bin/custom/tyreshopcore/src/com/reply/tyreshop/core/dto/ExchangeDto.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    private String cur_Name;
+
+    @JsonProperty("Cur_OfficialRate")
+    private Double cur_OfficialRate;
+
+    public ExchangeDto() {
+    }
+
+    public ExchangeDto(Integer cur_Id, Date date, String cur_Abbreviation, Integer cur_Scale, String cur_Name,
+                       Double cur_OfficialRate) {
```
Comment (uhpavel) [inline bin/custom/tyreshopcore/src/com/reply/tyreshop/core/dto/ExchangeDto.java]: do we need this constructor

File: bin/custom/tyreshopcore/src/com/reply/tyreshop/core/job/TyreshopUpdateCurrencyRatesJob.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+import javax.annotation.Resource;
+import java.util.List;
+
+public class TyreshopUpdateCurrencyRatesJob extends AbstractJobPerformable<CronJobModel> {
+
+    private static final Logger LOG = Logger.getLogger(TyreshopUpdateCurrencyRatesJob.class);
+
+    @Resource
+    ModelService modelService;
```
Comment (uhpavel) [inline bin/custom/tyreshopcore/src/com/reply/tyreshop/core/job/TyreshopUpdateCurrencyRatesJob.java]: use modelservice from parent class

File: bin/custom/tyreshopcore/src/com/reply/tyreshop/core/job/TyreshopUpdateCurrencyRatesJob.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    @Override
+    public PerformResult perform(CronJobModel cronJobModel) {
+        boolean success = true;
+        List<CurrencyModel> currencies = currencyDao.findCurrencies();
+        for (CurrencyModel currency : currencies){
+            if(!currency.getIsocode().equals(
+                    configurationService.getConfiguration().getString("exchangerates.currency"))){
+                try{
+                    currency.setConversion(currenciesRatesService.getExchangeRate(currency.getIsocode()));
+                    modelService.save(currency);
```
Comment (uhpavel) [inline bin/custom/tyreshopcore/src/com/reply/tyreshop/core/job/TyreshopUpdateCurrencyRatesJob.java]: create temp list with all modified objects and then use method saveAll

File: bin/custom/tyreshopcore/src/com/reply/tyreshop/core/job/TyreshopUpdateCurrencyRatesJob.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+                }
+                catch (RestClientException ex){
```
Comment (uhpavel) [inline bin/custom/tyreshopcore/src/com/reply/tyreshop/core/job/TyreshopUpdateCurrencyRatesJob.java]: Catch this exception in exchange rate service and catch here custom exception e.g. ExchangeRateRetrievalException

File: bin/custom/tyreshopcore/src/com/reply/tyreshop/core/services/impl/DefaultTyreshopExchangeRateService.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+    private ConfigurationService configurationService;
+    private RestOperations restOperations;
+
+    @Override
+    public ExchangeDTO[] getExchangeRates() throws ExchangeRateRetrievalException {
+        String uri = configurationService.getConfiguration().getString("exchangerates.uri");
+        ExchangeDTO[] exchangeDTOs;
+        try {
+            exchangeDTOs = restOperations.getForObject(uri, ExchangeDTO[].class);
```
Comment (uhpavel) [inline bin/custom/tyreshopcore/src/com/reply/tyreshop/core/services/impl/DefaultTyreshopExchangeRateService.java]: Change array to list

Comment (uhpavel) [reply bin/custom/tyreshopcore/src/com/reply/tyreshop/core/services/impl/DefaultTyreshopExchangeRateService.java]: even better to return here map<currency key, exchangerate value>

File: bin/custom/tyreshopcore/src/com/reply/tyreshop/core/job/TyreshopUpdateCurrencyRatesJob.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Override
+    public PerformResult perform(CronJobModel cronJobModel) {
+        Map<String, Double> exchangeRates;
+        try {
+            exchangeRates = exchangeRateService.getExchangeRates();
+        }
+        catch (ExchangeRateRetrievalException ex){
+            LOG.error("Error receiving exchange rates!");
+            return new PerformResult(CronJobResult.FAILURE, CronJobStatus.FINISHED);
```
Comment (andy-kru) [inline bin/custom/tyreshopcore/src/com/reply/tyreshop/core/job/TyreshopUpdateCurrencyRatesJob.java]: change CronJobStatus to CronJobStatus.ABORTED.

CronJobStatus.FINISHED:
The CronJob has completed its run.
CronJobStatus.ABORTED:
The CronJob's execution has been cancelled - either due to an error or due to user intervention. Check the CronJob's log for details.

File: bin/custom/tyreshopcore/src/com/reply/tyreshop/core/job/TyreshopUpdateCurrencyRatesJob.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    private static final Logger LOG = Logger.getLogger(TyreshopUpdateCurrencyRatesJob.class);
+
+    private CurrencyDao currencyDao;
+    private TyreshopExchangeRateService exchangeRateService;
+
+    @Override
+    public PerformResult perform(CronJobModel cronJobModel) {
+        Map<String, Double> exchangeRates;
+        try {
+            exchangeRates = exchangeRateService.getExchangeRates();
```
Comment (andy-kru) [inline bin/custom/tyreshopcore/src/com/reply/tyreshop/core/job/TyreshopUpdateCurrencyRatesJob.java]: Use getters for all injected fields

File: bin/custom/tyreshopcore/src/com/reply/tyreshop/core/services/impl/DefaultTyreshopExchangeRateService.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.stream.Collectors;
+
+public class DefaultTyreshopExchangeRateService implements TyreshopExchangeRateService {
+
+    private ConfigurationService configurationService;
+    private RestOperations restOperations;
+
+    @Override
+    public Map<String, Double> getExchangeRates() throws ExchangeRateRetrievalException {
+        String uri = configurationService.getConfiguration().getString("exchangerates.uri");
```
Comment (andy-kru) [inline bin/custom/tyreshopcore/src/com/reply/tyreshop/core/services/impl/DefaultTyreshopExchangeRateService.java]: Use getters for all injected fields

File: bin/custom/tyreshopcore/src/com/reply/tyreshop/core/services/impl/DefaultTyreshopExchangeRateService.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+        catch (RestClientException ex){
+            throw new ExchangeRateRetrievalException();
+        }
+        Collection<ExchangeDTO> exchangeDTOCollection = response.getBody();
+        Map<String, Double> exchangeRates = exchangeDTOCollection.stream().collect(
+                Collectors.toMap(ExchangeDTO::getCurAbbreviation, ExchangeDTO::getCurOfficialRate));
+        return exchangeRates;
+    }
+
+    public void setConfigurationService(ConfigurationService configurationService) {
```
Comment (andy-kru) [inline bin/custom/tyreshopcore/src/com/reply/tyreshop/core/services/impl/DefaultTyreshopExchangeRateService.java]: You can mark mandatory fields with @Required annotation

File: bin/custom/tyreshopcore/src/com/reply/tyreshop/core/job/TyreshopUpdateCurrencyRatesJob.java [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+            if(exchangeRates.containsKey(currencyModel.getIsocode())){
+                currencyModel.setConversion(exchangeRates.get(currencyModel.getIsocode()));
+                updated.add(currencyModel);
+            }
+        }
+        modelService.saveAll(updated);
+        return new PerformResult(CronJobResult.SUCCESS, CronJobStatus.FINISHED);
+    }
+
+    public void setCurrencyDao(CurrencyDao currencyDao) {
```
Comment (andy-kru) [inline bin/custom/tyreshopcore/src/com/reply/tyreshop/core/job/TyreshopUpdateCurrencyRatesJob.java]: You can mark mandatory fields with @Required annotation

**Pull request merged**